runtime/krest/src/lib.rs

```diff
@@ -428,20 +436,104 @@ impl WeightToFeePolynomial for WeightToFee {
-				tips.merge_into(&mut fees);
-			}
-			// pay fees to collators
-			<ToStakingPot as OnUnbalanced<_>>::on_unbalanced(fees);
-		}
+	// Overwrite on_unbalanced() and on_nonzero_unbalanced(), because their default
+	// implementations will just drop the imbalances!! Instead on_unbalanceds() will
+	// use these two following methods.
+	fn on_unbalanced(amount: NegativeImbalance) {
+		Self::on_nonzero_unbalanced(amount);
+	}
+
+	fn on_nonzero_unbalanced(amount: NegativeImbalance) {
+		<BlockReward as OnUnbalanced<_>>::on_unbalanced(amount);
 	}
 }
 
+type NegativeImbalanceOf<C, T> =
+	<C as Currency<<T as frame_system::Config>::AccountId>>::NegativeImbalance;
+
+pub struct PeaqCurrencyAdapter<C, OU>(PhantomData<(C, OU)>);
+
+impl<T, C, OU> OnChargeTransaction<T> for PeaqCurrencyAdapter<C, OU>
+where
+    T: TransactionPaymentConfig,
+    C: Currency<<T as frame_system::Config>::AccountId>,
+    OU: OnUnbalanced<NegativeImbalanceOf<C, T>>,
+{
+    type LiquidityInfo = Option<NegativeImbalanceOf<C, T>>;
+    type Balance = <C as Currency<<T as frame_system::Config>::AccountId>>::Balance;
+
+    /// Withdraw the predicted fee from the transaction origin.
+    /// Note: The `fee` already includes the `tip`.
+    fn withdraw_fee(
+        who: &T::AccountId,
+        _call: &T::Call,
+        _info: &DispatchInfoOf<T::Call>,
+        fee: Self::Balance,
+        tip: Self::Balance,
+    ) -> Result<Self::LiquidityInfo, TransactionValidityError> {
+		let network_fee = fee;
+        if network_fee.is_zero() {
+            return Ok(None)
+        }
+
+        let withdraw_reason = if tip.is_zero() {
+            WithdrawReasons::TRANSACTION_PAYMENT
+        } else {
+            WithdrawReasons::TRANSACTION_PAYMENT | WithdrawReasons::TIP
+        };
+
+		// Apply Peaq Economy-of-Things Fee adjustment
+		let reward_fee = Perbill::from_percent(50) * network_fee;
+        let tx_fee = network_fee.saturating_add(reward_fee);
+
+        match C::withdraw(who, tx_fee, withdraw_reason, ExistenceRequirement::KeepAlive) {
+            Ok(imbalance) => Ok(Some(imbalance)),
+            Err(_) => Err(InvalidTransaction::Payment.into()),
+        }
+    }
+
+    /// Hand the fee and the tip over to the `[OnUnbalanced]` implementation.
+    /// Since the predicted fee might have been too high, parts of the fee may
+    /// be refunded.
+    /// Note: The `corrected_fee` already includes the `tip`.
+    fn correct_and_deposit_fee(
+        who: &T::AccountId,
+        _dispatch_info: &DispatchInfoOf<T::Call>,
+        _post_info: &PostDispatchInfoOf<T::Call>,
+        corrected_fee: Self::Balance,
+        tip: Self::Balance,
+        already_withdrawn: Self::LiquidityInfo,
+    ) -> Result<(), TransactionValidityError> {
+        if let Some(paid) = already_withdrawn {
+			// Apply same Peaq Economy-of-Things Fee adjustment as above
+			let cor_network_fee = corrected_fee;
+			let cor_reward_fee = Perbill::from_percent(50) * corrected_fee;
```

Comment:
It would be easy to extract it as a constant into our runtime. I think moving it as a configurable parameter within the block-reward pallet would be not a good solution, because currently this factor is implemented within an separat adapter for the transaction-payment-pallet... would be a strange connection between those 3 components...

If we define it as a constant, it would be a little nicer :+1: Will do it...

Reply:
Okay, then, let us use the constant first. And once the other calculation comes up, we must consider some strategy patterns to improve it.